### PR TITLE
Always copy GameState passed into think or ponder (fix only no gtest)

### PR DESCRIPTION
**Pull request by killerducky**
Fix only version of #711.

Might take awhile to get Travis working so here is the bug fix only.

### CODE REVIEW

File: src/UCTSearch.h [hunk 1]
```diff
@@ -76,6 +76,7 @@ class UCTSearch {
 
     UCTSearch();
     void set_gamestate(const GameState& g);
+    GameState get_gamestate() const;
```
Comment (gcp) [inline src/UCTSearch.h]: You aren't using this as far as I can see? (And if you were, shouldn't it return a ref?)

Comment (killerducky) [reply src/UCTSearch.h]: It's used by the gtest version. I will remove it from this pull and have it only in the gtest pull version.

**Comment by roy7** [comment]
Love you all if we can finish this one up. :)

**Comment by killerducky** [comment]
I couldn't get it to work. Relevant code:
```
void UCTNode::find_new_root(const GameState& g_new,
                                  GameState& g_curr) {
    node_ptr_t root = nullptr;
    root = std::move(child);
    *this = root;


/home/aolsen/projects/lz-tc/leela-zero/src/UCTNode.cpp: In member function ‘void UCTNode::find_new_root(const GameState&, GameState&)’:
/home/aolsen/projects/lz-tc/leela-zero/src/UCTNode.cpp:454:17: error: no match for ‘operator=’ (operand types are ‘UCTNode’ and ‘UCTNode::node_ptr_t {aka std::unique_ptr<UCTNode>}’)
         *this = root;
```

**Comment by killerducky** [comment]
This is ready, unless someone can help me get an alternate way of doing find_new_root working. My Google-fu for "std::move this" didn't find anything helpful.

**Comment by bood** [comment]
#725 is almost ready now, you can try your unit test code again after it's merged.

**Comment by gcp** [comment]
I tried to solve the f->x(f, ...) issue by making find_new_root part of UCTSearch, which actually simplifies things a lot, but then you run into the problem that it fundamentally needs to alter the children of an UCTNode, so it can't do that with the const ref get_children() returns. And it's a sign that UCTSearch shouldn't doing that kind of mucking around too.

File: src/UCTNode.cpp [hunk 2]
```diff
@@ -424,26 +424,32 @@ UCTNode::node_ptr_t UCTNode::find_new_root(const int move) {
...
+                        root = std::move(child);
+                        break;
+                    }
+                    g_curr.undo_move();
                 }
-                g_curr.undo_move();
             }
         }
+        // No match. Create a new root.
+        root = std::make_unique<UCTNode>(FastBoard::PASS, 0.0f, 0.5f);
```
Comment (gcp) [inline src/UCTNode.cpp]: Isn't this in the wrong place now? If komi or hash mismatches, the root won't be reset. If a move matches, we'll break out of the for and overwrite here?

**Comment by gcp** [comment]
I reworked this a bit. I can't say why, but I prefer x = x->f(...) over x->f(x, ...).

It also makes the signature and usage pattern of both calls match.

File: src/UCTNode.cpp (hunk 2 above)
Comment (killerducky) [reply src/UCTNode.cpp]: Oops good catch. I would have found this if I retested self-play. I'll work on a gtest for these cases later.

**Pull request merged**